src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -79,6 +80,8 @@ public virtual IDisposable Start()
                         contextAccessor.HttpContext = httpContext;
                         await application(httpContext);
                     }
+
+                    return httpContext;
```

Comment:
I'm not recommending any specific solution, just trying to outline what's possible. HttpContext is supported in OnCompleted, but more importantly it's supported in OnStarted which can fire after `await application(httpContext)` in no-body scenarios.